File: bundle/edu.gemini.util.skycalc/src/main/java/edu/gemini/skycalc/Offset.java
//
// $
//

package edu.gemini.skycalc;

/**
 * Offset coordinates expressed as angular separations between two points, a
 * base position and a second position.
 */
public final class Offset {
    /** Min value for two offsets to be different in arc seconds; this value should be used for comparisons of offsets. */
    private static final double MIN_VALUE = 0.000001;

    private static final Angle ZERO_ARCSECS = new Angle(0, Angle.Unit.ARCSECS);

    /**
     * An Offset of (0, 0) arcsecs.
     */
    public static final Offset ZERO_OFFSET = new Offset(ZERO_ARCSECS, ZERO_ARCSECS);

    private final Angle p;
    private final Angle q;

    public Offset(Angle p, Angle q) {
        if (p == null) throw new NullPointerException("p is null");
        if (q == null) throw new NullPointerException("q is null");

        this.p = p;
        this.q = q;
    }

    public Angle p() {
        return p;
    }

    public Angle q() {
        return q;
    }

    /**
     * Calculates the distance between the base position and the offset
     * position, expressed in the units of p.
     *
     * @return angular separation between base position and offset position
     */
    public Angle distance() {
        return distance(Offset.ZERO_OFFSET);
    }

    /**
     * Calculates the absolute distance between this Offset position and the
     * given position, expressed in the units of this offset's p.
     *
     * @param other distance computed is relative to the given offset
     *
     * @return distance, expressed in the units of this offset's p, between
     * this offset and the given <code>other</code> offset
     */
    public Angle distance(Offset other) {
        Angle.Unit unit = p.getUnit();

        double thisP = p.getMagnitude();
        double thisQ = q.convertTo(unit).getMagnitude();

        double thatP = other.p().convertTo(unit).getMagnitude();
        double thatQ = other.q().convertTo(unit).getMagnitude();

        double p = thisP - thatP;
        double q = thisQ - thatQ;

        double d = Math.sqrt(p*p + q*q);
        return new Angle(d, unit);
    }

    /**
     * Checks if the size of an offset value in arcseconds can be considered to be zero.
     * Helper method to improve legibility.
     * @param a
     * @return
     */
    public static boolean isZero(final double a) { return Math.abs(a) < MIN_VALUE; }

    /**
     * Checks if two offset values in arcseconds can be considered equal.
     * Helper method to improve legibility.
     * @param a
     * @param b
     * @return
     */
    public static boolean areEqual(final double a, final double b) { return Math.abs(a - b) < MIN_VALUE; }

    public String toString() {
        return String.format("%f x %f", p.getMagnitude(), q.getMagnitude());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Offset that = (Offset) o;

        if (!p.equals(that.p)) return false;
        return q.equals(that.q);
    }

    @Override
    public int hashCode() {
        int result = p.hashCode();
        result = 31 * result + q.hashCode();
        return result;
    }
}

File: bundle/edu.gemini.spModel.core/src/main/scala/edu/gemini/spModel/core/Offset.scala
package edu.gemini.spModel.core

import AngleSyntax._

import scala.math._

import scalaz._
import Scalaz._

/**
 * Offset coordinates expressed as angular separations between two points, a
 * base position and a second position.
 */
case class Offset(p: OffsetP, q: OffsetQ) {

  /**
   * Calculates the absolute distance between this Offset position and the
   * given position
   *
   * @param other distance computed is relative to the given offset
   *
   * @return distance, expressed as an Angle, between
   *         this offset and the given <code>other</code> offset
   */
  def distance(other: Offset): Angle = {
    val pd = p.degrees - other.p.degrees
    val qd = q.degrees - other.q.degrees
    val d = sqrt(pd*pd + qd*qd)

    Angle.fromDegrees(d)
  }

  /**
   * Calculates the bearing from 0,0 as an Angle clockwise from north 0º.
   */
  def bearing: Angle = {
    val x = -p.toAngle.toSignedDegrees
    val y =  q.toAngle.toSignedDegrees
    Angle.fromRadians(atan2(x, y))
  }

  /**
   * Calculates the distance between the zero position and the offset
   * position
   *
   * @return angular separation between base position and offset position
   */
  def distance: Angle = distance(Offset.zero)

  def +(that: Offset): Offset = Offset(p + that.p, q + that.q)
  def -(that: Offset): Offset = Offset(p - that.p, q - that.q)

  def *(scale: Double): Offset = Offset(p * scale, q * scale)

  def flipP: Offset = Offset(p * -1, q)
  def flipQ: Offset = Offset(p, q * -1)
}

object Offset {
  val zero = Offset(OffsetP.Zero, OffsetQ.Zero)

  /** @group Typeclass Instances */
  implicit val equals = scalaz.Equal.equalA[Offset]

  /** @group Typeclass Instances. */
  implicit val MonoidOffset: Monoid[Offset] =
    new Monoid[Offset] {
      val zero = Offset.zero
      def append(a: Offset, b: => Offset): Offset = a + b
    }

  implicit val ShowOffset: Show[Offset] =
    Show.shows[Offset] { off => s"Offset(${off.p.shows}, ${off.q.shows})" }
}